fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1062,6 +1055,315 @@ Future<Void> managementClusterRemoveTenantFromGroup(Transaction tr,
+			           self->ctx.runManagementTransaction([self = self](Reference<typename DB::TransactionT> tr) {
+				           return markClusterAsRestoring(self, tr);
+			           })));
+		} catch (Error& e) {
+			// If the transaction retries after success or if we are trying a second time to restore the cluster, it
+			// will throw an error indicating that the restore has already started
+			if (e.code() == error_code_cluster_restored) {
+				clusterIsRestoring = true;
+			} else {
+				throw;
+			}
+		}
+
+		if (!clusterIsRestoring) {
+			// get all the tenant information from the newly registered data cluster
+			wait(getAllTenantsFromDataCluster(self));
+			// get all the tenant information for this data cluster from manangement cluster
+			wait(getAllTenantsFromManagementCluster(self));
+
+			state std::unordered_map<TenantName, TenantMapEntry>::iterator itr = self->dataClusterTenantMap.begin();
+			while (itr != self->dataClusterTenantMap.end()) {
+
+				state TenantName tenantNameOnDataCluster = itr->first;
+				state uint64_t tenantId = (itr->second).id;
+				self->tenantId = tenantId;
+				self->tenantName = tenantNameOnDataCluster;
+
+				state Optional<TenantName> tenantNameOnMgmtCluster = wait(lookupTenantInManagementClusterById(self));
+				if (!tenantNameOnMgmtCluster.present()) {
+					// A tenant with this id is not found on the mgmt cluster.
+					if (self->removeMissingTenants == RemoveMissingTenants::True) {
+						wait(removeTenantFromDataCluster(self));
+					}
+				} else {
+					// A tenant with this id is found on the manangement cluster.
+					if (tenantNameOnDataCluster.compare(tenantNameOnMgmtCluster.get()) != 0) {
+						// renamed
+						self->tenantNewName = tenantNameOnMgmtCluster.get();
+						wait(renameTenantOnDataCluster(self));
+					}
+				}
+
+				++itr;
+			}
+
+			state std::unordered_set<TenantNameRef>::iterator setItr = self->mgmtTenantSet.begin();
+			while (setItr != self->mgmtTenantSet.end()) {
+				self->tenantId = self->mgmtClusterTenantMap[*setItr].id;
+				state Optional<TenantName> tNameOnDataCluster = wait(lookupTenantInDataClusterById(self));
+
+				if (!tNameOnDataCluster.present()) {
+					// Set Tenant in ERROR state
+					wait(self->ctx.runManagementTransaction([self = self](Reference<typename DB::TransactionT> tr) {
+						return markManagementTenantAsError(self, tr);
+					}));
+				}
+				++setItr;
+			}
+
+			// set restored cluster to ready state
+			try {
+				wait(self->ctx.runManagementTransaction(
+				    [self = self](Reference<typename DB::TransactionT> tr) { return markClusterAsReady(self, tr); }));
+			} catch (Error& e) {
+				throw;
+			}
+		}
+
+		return Void();
+	}
+
+	Future<Void> run() { return run(this); }
+};
+
+ACTOR template <class DB>
+Future<Void> restoreCluster(Reference<DB> db,
+                            ClusterName name,
+                            ClusterConnectionString connectionString,
+                            DataClusterEntry entry,
```

Comment:
I suppose `DataClusterEntry` should be an optional, since in the case where we are restoring an existing data cluster we shouldn't be providing one (we can get it from the management cluster).

In the case where we are restoring the management cluster, we could provide it here (meaning that we need it to be possible for EFDB to specify the parameters again), or we need to derive it from the data cluster. We can worry about that part later.